### DISCUSSION TITLE
How do subscriptions use @Root?

### DISCUSSION THREAD

**Discussion by gblikas** [MichalLytek/type-graphql#1212] [Q&A] [answered]
In https://github.com/MichalLytek/type-graphql/blob/master/examples/redis-subscriptions/recipe.resolver.ts, the `newComments` method has a `@Root` parameter, https://github.com/MichalLytek/type-graphql/blob/4fc77f542f27dd8b9c439c93cc9cf10a9bdc6469/examples/redis-subscriptions/recipe.resolver.ts#L56-L71 

I'm having a hard time understanding _how_ a method decorated with `@Subscription` can have a root object; calling `newComments` as a subscription is - as in the examples, https://github.com/MichalLytek/type-graphql/blob/4fc77f542f27dd8b9c439c93cc9cf10a9bdc6469/examples/redis-subscriptions/examples.gql#L29-L35 This implies no higher-order object, whereas, using `@Root` seems to imply that you could do something like 

```graphql
someCommentOnARecipieThing {
  newComments(recipeId: "2") {
      nickname
      content
      date
    }
}
```

Can I get an example of _how_ `type-graphql` uses a `@Root` object in relation to a subscription? Perhaps, how can I re-write my last code-block to be "correct"?

> I feel like I'm not understanding the nuance between how `type-graphql` uses `topics`, `filters` and `subscriptions`, when contrasted against a more-traditional implementation that might use a `subscribe: () => pubsub.asyncIterator(["SOME_TOPIC"])` pattern.

In contrast, the `@Args() { recipeId }: NewCommentsArgs` part makes total sense, in conjunction with the `@subscription` decorator; you want to subscribe to some recipe's new comments, in context `recipeId: "2"`.

**Comment by MichalLytek** [comment, marked as answer]
I did that 4 years ago, I wouldn't name that now `@Root()` 😅 
Some GraphQL tools call it parent, source, root - it's basically the first parameter of the resolver.

In case of subscriptions, it's the payload from the pub sub - the published message.
